rclcpp/include/rclcpp/executor.hpp

```diff
@@ -350,6 +351,51 @@ class Executor
   virtual void
   spin_once(std::chrono::nanoseconds timeout = std::chrono::nanoseconds(-1));
 
+  /// Spin (blocking) until the condition is complete, it times out waiting,
+  /// or rclcpp is interrupted.
+  /**
+   * \param[in] future The condition which can be callable or future type to wait on.
+   *   If this function returns SUCCESS, the future can be
+   *   accessed without blocking (though it may still throw an exception).
+   * \param[in] timeout Optional timeout parameter, which gets passed to Executor::spin_node_once.
+   *   `-1` is block forever, `0` is non-blocking.
+   *   If the time spent inside the blocking loop exceeds this timeout, return a TIMEOUT return
+   *   code.
+   * \return The return code, one of `SUCCESS`, `INTERRUPTED`, or `TIMEOUT`.
+   */
+  template<typename ConditionT, typename DurationT = std::chrono::milliseconds>
+  FutureReturnCode
+  spin_until_complete(
+    const ConditionT & condition,
+    DurationT timeout = DurationT(-1))
```

Comment:
Any suggestion? We could add a more generic version of `FutureReturnCode` (what name?), but it would be functionally the same.

Reply:
That's a good step, but I was also thinking, since the Condition must return a bool, should this function not return one? The reason for a complex type for futures is that it is multi-state, but that doesn't seem to be true for the condition.

Reply:
Can `spin_until_complete` not be `INTERRUPTED` even for non-future conditions? It just means it stopped spinning because `rclcpp::ok()` returned `false`.

Reply:
That's true, I suppose.

With that in mind, the name `rclcpp::ConditionWaitReturnCode` is kind of awkward, since it's the first time "Wait" is used in this context (the function isn't called "spin_and_wait_for_condition()" or something like that), but I'm not sure of an alternative, maybe the explicit, but verbose, `rclcpp::SpinUntilCompleteReturnCode`?

Another route would be to copy `std::future_status` (https://en.cppreference.com/w/cpp/thread/future_status), and call it something like `rclcpp::condition_status`, but that (over)uses the term `condition` in my opinion, and we might want that later, so I'd prefer something more namespaced, like `rclcpp::SpinUntilCompleteStatus`.

I'll leave it up to you, but I like the `rclcpp::SpinUntilCompleteReturnCode` option, since it matches our prexisting `rclcpp::FutureReturnCode`.

Reply:
Sounds good, I changed it to `rclcpp::SpinUntilCompleteReturnCode`.